### PR TITLE
Live waveform

### PR DESCRIPTION
**Pull request by langleyd**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Show a live waveform when the user is recording a voice message

## Screenshots / GIFs

https://github.com/vector-im/element-x-android/assets/6135282/a98ba4d4-8f13-488a-be92-e79c65949cec

### CODE REVIEW

File: tests/uitests/src/test/snapshots/images/ui_S_t[l.textcomposer_null_TextComposerVoice-D-4_4_null,NEXUS_5,1.0,en].png
Comment (langleyd) [inline tests/uitests/src/test/snapshots/images/ui_S_t[l.textcomposer_null_TextComposerVoice-D-4_4_null,NEXUS_5,1.0,en].png:1]: Hrmm that doesn't seem right, will take a look.

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/composer/VoiceMessageComposerStateProvider.kt [hunk 1]
```diff
@@ -35,3 +36,7 @@ internal fun aVoiceMessageComposerState(
     showPermissionRationaleDialog = showPermissionRationaleDialog,
     eventSink = {},
 )
+
+internal var aWaveformLevels = List(100) { it.toFloat() / 200 }.toPersistentList()
```
Comment (julioromano) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/composer/VoiceMessageComposerStateProvider.kt:40]: `/ 200` won't show full scale (will reach 0.5 only). Is this intended?

File: libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/TextComposer.kt [hunk 2]
```diff
@@ -798,7 +799,7 @@ internal fun TextComposerVoicePreview() = ElementPreview {
         enableVoiceMessages = true,
     )
     PreviewColumn(items = persistentListOf({
-        VoicePreview(voiceMessageState = VoiceMessageState.Recording(61.seconds, 0.5f))
+        VoicePreview(voiceMessageState = VoiceMessageState.Recording(61.seconds, List(100) { it.toFloat() / 200 }.toPersistentList()))
```
Comment (julioromano) [inline libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/TextComposer.kt:802]: Same here, perhaps showing from 0 to full scale is better in screenshots?

File: libraries/voicerecorder/impl/src/main/kotlin/io/element/android/libraries/voicerecorder/impl/VoiceRecorderImpl.kt [hunk 3]
```diff
@@ -67,6 +69,7 @@ class VoiceRecorderImpl @Inject constructor(
     private var audioReader: AudioReader? = null
     private var recordingJob: Job? = null
     private val levels: MutableList<Float> = mutableListOf()
+    private val lock = Mutex()
```
Comment (julioromano) [inline libraries/voicerecorder/impl/src/main/kotlin/io/element/android/libraries/voicerecorder/impl/VoiceRecorderImpl.kt:72]: Was there some concurrency thing that led to crashes before?

**Review by julioromano — APPROVED** [review]
LGMT, just some nits

File: libraries/voicerecorder/impl/src/main/kotlin/io/element/android/libraries/voicerecorder/impl/VoiceRecorderImpl.kt (hunk 3 above)
Comment (langleyd) [reply libraries/voicerecorder/impl/src/main/kotlin/io/element/android/libraries/voicerecorder/impl/VoiceRecorderImpl.kt:72]: Yes a concurrency modification exception when calling toPersisentList() at a later point, as I was returning a list that was being modified, so some safety measures.

**Pull request merged**